File: android/src/main/java/com/reactnativespacex/impl/transport/CommonTransportLayer.kt
package com.reactnativespacex.impl.transport

import android.annotation.SuppressLint
import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.os.Build
import com.reactnativespacex.impl.SpaceXMode
import com.reactnativespacex.impl.transport.net.ThrustedSocket
import com.reactnativespacex.impl.utils.DispatchQueue
import com.reactnativespacex.impl.utils.fatalError
import com.reactnativespacex.impl.utils.xLog
import org.json.JSONObject

enum class NetworkingApolloState {
  WAITING, CONNECTING, STARTING, STARTED, COMPLETED
}

val PING_INTERVAL = 1000
val SOCKET_TIMEOUT = 5000

@SuppressLint("MissingPermission")
class CommonTransportLayer(
  val context: Context,
  val url: String,
  val params: Map<String, String?>,
  val mode: SpaceXMode,
  private val handlerQueue: DispatchQueue,
  private val handler: NetworkingHandler
) {

  private val queue = DispatchQueue("spacex-networking-apollo")
  private var state = NetworkingApolloState.WAITING
  private var client: ThrustedSocket? = null
  private var failuresCount = 0
  private var reachable = false
  private var started = false
  private var pending = mutableMapOf<String, JSONObject>()
  private var pingTimeout: (() -> Unit)? = null

  private val connectivityCallback = object : ConnectivityManager.NetworkCallback() {
    override fun onAvailable(network: Network) {
      queue.async {
        onReachable()
      }
    }

    override fun onUnavailable() {
      queue.async {
        onUnreachable()
      }
    }

    override fun onLost(network: Network) {
      queue.async {
        onUnreachable()
      }
    }
  }

  init {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
      connectivityManager.registerDefaultNetworkCallback(connectivityCallback)
    } else {
      // Enforce network for old androids
      reachable = true
    }
  }

  fun connect() {
    queue.async {
      xLog("SpaceX-Apollo", "Starting")
      this.started = true
      if (this.reachable) {
        doConnect()
      }
    }
  }

  fun startRequest(id: String, body: JSONObject) {
    queue.async {
      if (state == NetworkingApolloState.WAITING || state == NetworkingApolloState.CONNECTING) {
        // Add to pending buffer if we are not connected already
        this.pending[id] = body
      } else if (state == NetworkingApolloState.STARTING) {
        // If we connected, but not started add to pending buffer (in case of failed start)
        // and send message to socket

        this.pending[id] = body
        this.writeToSocket(JSONObject(mapOf(
          "type" to "start",
          "id" to id,
          "payload" to body
        )))
      } else if (state == NetworkingApolloState.STARTED) {
        this.writeToSocket(JSONObject(mapOf(
          "type" to "start",
          "id" to id,
          "payload" to body
        )))
      } else if (state == NetworkingApolloState.COMPLETED) {
        // Silently ignore if connection is completed
      } else {
        fatalError()
      }
    }
  }

  fun stopRequest(id: String) {
    queue.async {
      if (state == NetworkingApolloState.WAITING || state == NetworkingApolloState.CONNECTING) {
        // Remove from pending buffer if we are not connected already
        this.pending.remove(id)
      } else if (state == NetworkingApolloState.STARTING) {
        // If we connected, but not started remove from pending buffer (in case of failed start)
        // and send cancellation message to socket
        this.pending.remove(id)
        this.writeToSocket(JSONObject(mapOf("type" to "stop", "id" to id)))
      } else if (state == NetworkingApolloState.STARTED) {
        this.writeToSocket(JSONObject(mapOf("type" to "stop", "id" to id)))
      } else if (state == NetworkingApolloState.COMPLETED) {
        // Silently ignore if connection is completed
      } else {
        fatalError()
      }
    }
  }

  fun close() {
    queue.async {
      if (this.state !== NetworkingApolloState.COMPLETED) {
        this.state = NetworkingApolloState.COMPLETED

        // Remove all pending requests
        this.pending.clear()

        // Stopping client
        this.stopClient()

        // Stopping network state callback
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
          val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
          connectivityManager.unregisterNetworkCallback(connectivityCallback)
        }
      }
    }
  }

  private fun doConnect() {
    xLog("SpaceX-Transport", "Connecting")
    if (this.state != NetworkingApolloState.WAITING) {
      fatalError("Unexpected state")
    }
    this.state = NetworkingApolloState.CONNECTING

    val ws = ThrustedSocket(this.url,
      if (this.mode == SpaceXMode.OPENLAND) null else "graphql-ws",
      SOCKET_TIMEOUT,
      this.queue
    )
    ws.onOpen = {
      if (this.client == ws) {
        onConnected()
      }
    }
    ws.onClose = {
      if (this.client == ws) {
        onDisconnected()
      }
    }
    ws.onMessage = {
      if (this.client == ws) {
        onReceived(it)
      }
    }
    this.client = ws
  }

  private fun onConnected() {
    xLog("SpaceX-Transport", "Connected")
    if (this.state != NetworkingApolloState.CONNECTING) {
      fatalError("Unexpected state")
    }
    this.state = NetworkingApolloState.STARTING

    if (this.mode == SpaceXMode.OPENLAND) {
      this.writeToSocket(JSONObject(mapOf(
        "protocol_v" to 2,
        "type" to "connection_init",
        "payload" to this.params
      )))
    } else {
      this.writeToSocket(JSONObject(mapOf(
        "type" to "connection_init",
        "payload" to this.params
      )))
    }
    for (p in this.pending) {
      this.writeToSocket(JSONObject(mapOf(
        "type" to "start",
        "id" to p.key,
        "payload" to p.value
      )))
    }
  }

  private fun sendPing() {
    this.pingTimeout?.invoke()
    this.pingTimeout = this.queue.asyncDelayed(PING_INTERVAL) {
      if (this.mode == SpaceXMode.OPENLAND) {
        writeToSocket(JSONObject(mapOf(
          "type" to "ping"
        )))
      }
    }
  }

  private fun onReceived(message: String) {
    // xLog("SpaceX-Apollo", "<< $message")

    val parsed = JSONObject(message)
    val type = parsed.getString("type")
//        xLog("SpaceX-Apollo", "<< $type")
    if (type == "ka") {
      // Ignore
    } else if (type == "connection_ack") {
      if (this.state == NetworkingApolloState.STARTING) {
        xLog("SpaceX-Transport", "Started")

        // Change State
        this.state = NetworkingApolloState.STARTED

        // Remove all pending messages
        this.pending.clear()

        // Reset failure count
        this.failuresCount = 0

        // Ping
        sendPing()

        // Notify about state
        this.handlerQueue.async {
          this.handler.onConnected()
        }
      }
    } else if (type == "data") {
      val id = parsed.getString("id")
      val payload = parsed.getJSONObject("payload")
      val errors = payload.optJSONArray("errors")
      if (errors != null) {
        this.handlerQueue.async {
          xLog("SpaceX-Transport", "Error ($id)")
          this.handler.onError(id, errors)
        }
      } else {
        val data = payload.getJSONObject("data")
        this.handlerQueue.async {
          xLog("SpaceX-Transport", "Data ($id)")
          this.handler.onResponse(id, data)
        }
      }
    } else if (type == "error") {
      val id = parsed.getString("id")
//            this.handlerQueue.async {
//                xLog("SpaceX-Apollo", "Critical Error ($id): Retrying")
//                this.handler.onCompleted(id)
//            }
    } else if (type == "ping") {
      if (this.mode === SpaceXMode.OPENLAND) {
        this.writeToSocket(JSONObject(mapOf(
          "type" to "pong"
        )))
      }
    } else if (type == "pong") {
      sendPing()
    } else if (type == "complete") {
      val id = parsed.getString("id")
      this.handlerQueue.async {
        xLog("SpaceX-Transport", "Complete ($id)")
        this.handler.onCompleted(id)
      }
    }
  }

  private fun onReachable() {
    xLog("SpaceX-Transport", "Reachable")
    this.reachable = true
    if (this.started && (this.state == NetworkingApolloState.WAITING
        || this.state == NetworkingApolloState.CONNECTING)) {
      this.stopClient()
      this.state = NetworkingApolloState.WAITING
      this.failuresCount = 0
      this.doConnect()
    }
  }

  private fun onUnreachable() {
    xLog("SpaceX-Transport", "Unreachable")
    this.reachable = false
  }

  private fun onDisconnected() {
    xLog("SpaceX-Transport", "Disconnected")
    if (this.state == NetworkingApolloState.STARTED) {
      this.handlerQueue.async {
        this.handler.onDisconnected()
        this.handler.onSessionRestart()
      }
    }
    this.stopClient()
    this.state = NetworkingApolloState.WAITING
    this.failuresCount += 1
    if (this.reachable) {
      this.doConnect()
    }
    this.queue.asyncDelayed(2000) {
      if (this.state == NetworkingApolloState.WAITING && this.reachable) {
        this.doConnect()
      }
    }
  }

  private fun stopClient() {
    val ws = client
    client = null

    ws?.onMessage = null
    ws?.onClose = null
    ws?.onOpen = null
    ws?.close()

    this.pingTimeout?.invoke()
    this.pingTimeout = null
  }

  private fun writeToSocket(src: JSONObject) {
    val msg = src.toString()
    xLog("SpaceX-Apollo", ">> $msg")
    this.client!!.post(msg)
  }
}

File: ios/networking/CommonTransportLayer.swift
//
//  ApolloTransport.swift
//  openland
//
//  Created by Steve Kite on 5/13/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON
import Starscream

enum TransportSocketState {
  case waiting
  case connecting
  case starting
  case started
  case completed
}

protocol NetworkingDelegate: AnyObject {
  
  // Callbacks
  func onResponse(id: String, data: JSON)
  func onError(id: String, error: JSON)
  func onCompleted(id: String)
  
  // Session state
  func onSessiontRestart()
  func onConnected()
  func onDisconnected()
}

class CommonTransportLayer: WebSocketConnectionDelegate {
  
  private static var nextId: AtomicInteger = AtomicInteger(value: 1)
  private static let PING_INTERVAL = 30
  private static let PING_TIMEOUT = 10
  
  private let queue = DispatchQueue(label: "spacex-networking-transport")
    
  let id: Int
  let url: String
  let params: [String: String?]
  weak var delegate: NetworkingDelegate? = nil
  var callbackQueue: DispatchQueue
  let provider: WebSocketProvider
  private var connection: WebSocketConnection? = nil
  private var pending: [String: JSON] = [:]
  private var state: TransportSocketState = .waiting
  private var failuresCount = 0
  private var started = false
  private var mode: String
  
  private var lastPingId = 0
  
  init(provider: WebSocketProvider, url: String, mode: String, params: [String: String?]) {
    self.callbackQueue = self.queue
    self.provider = provider
    self.id = CommonTransportLayer.nextId.getAndIncrement()
    self.url = url
    self.params = params
    self.mode = mode
  }
  
  func connect() {
    NSLog("[SpaceX-WS]: Starting")
    queue.async {
      self.started = true
      self.doConnect()
    }
  }
  
  func startRequest(id: String, body: JSON) {
    NSLog("[SpaceX-WS]: Start Request " + id + " [" + body["name"].stringValue + "]")
    queue.async {
      if self.state == .waiting || self.state == .connecting {
        // Add to pending buffer if we are not connected already
        self.pending[id] = body
        NSLog("[SpaceX-WS]: Pending " + id)
      } else if self.state == .starting {
        NSLog("[SpaceX-WS]: Starting " + id)
        // If we connected, but not started add to pending buffer (in case of failed start)
        // and send message to socket
        
        self.pending[id] = body
        self.writeToSocket(msg: JSON(["type": "start", "id": id, "payload": body]))
      } else if self.state == .started {
        NSLog("[SpaceX-WS]: Started " + id)
        self.writeToSocket(msg: JSON(["type": "start", "id": id, "payload": body]))
      } else if self.state == .completed {
        NSLog("[SpaceX-WS]: Completed " + id)
        // Silently ignore if connection is completed
      } else {
        fatalError()
      }
    }
  }
  
  func stopRequest(id: String) {
    NSLog("[SpaceX-WS]: Stop Request " + id)
    queue.async {
      if self.state == .waiting || self.state == .connecting {
        // Remove from pending buffer if we are not connected already
        self.pending.removeValue(forKey: id)
      } else if self.state == .starting {
        // If we connected, but not started remove from pending buffer (in case of failed start)
        // and send cancelation message to socket
        self.pending.removeValue(forKey: id)
        self.writeToSocket(msg: JSON(["type": "stop", "id": id]))
      } else if self.state == .started {
        self.writeToSocket(msg: JSON(["type": "stop", "id": id]))
      } else if self.state == .completed {
        // Silently ignore if connection is completed
      } else {
        fatalError()
      }
    }
  }
  
  private func doConnect() {
    NSLog("[SpaceX-WS]: Connecting")
    if self.state != .waiting {
      fatalError("Unexpected state")
    }
    self.state = .connecting
    
    // Create new connection
    var proto: String? = nil
    if (self.mode != "openland") {
        proto = "graphql-ws"
    }
    let ws = self.provider.create(endpoint: WebSocketEndpoint(url: self.url, proto: proto), queue: self.queue)
    ws.delegate = self
    self.connection = ws
  }
    
  func onOpen() {
    NSLog("[SpaceX-WS]: onConnected")
    if self.state != .connecting {
      fatalError("Unexpected state")
    }
    self.state = .starting
    if self.mode == "openland" {
      self.writeToSocket(msg: JSON([
        "protocol_v": 2,"type": "connection_init", "payload": self.params
      ]))
    } else {
      self.writeToSocket(msg: JSON([
        "type": "connection_init", "payload": self.params
      ]))
    }
    for p in self.pending {
      self.writeToSocket(msg: JSON(["type": "start", "id": p.key, "payload": p.value]))
    }
    schedulePing()
  }
    
  func onMessage(message: String) {
    let parsed = JSON(parseJSON: message)
    let type = parsed["type"].stringValue
    NSLog("[SpaceX-WS]: <<" + type)
    if type == "ka" {
      // TODO: Handle
    } else if type == "connection_ack" {
      if self.state == .starting {
        NSLog("[SpaceX-WS]: Started")
        
        // Change state
        self.state = .started
        
        // Remove all pending messages
        self.pending.removeAll()
        
        // Reset failure count
        self.failuresCount = 0
        
        // Notify about state
        self.callbackQueue.async {
          self.delegate?.onConnected()
        }
      }
    } else if type == "data" {
      let id = parsed["id"].stringValue
      let payload = parsed["payload"]
      let errors = payload["errors"]
      let data = payload["data"]
      if errors.exists() {
        self.callbackQueue.async {
            NSLog("[SpaceX-WS]: Error (" + id + ")")
            self.delegate?.onError(id: id, error: errors)
        }
      } else {
        NSLog("[SpaceX-WS]: Data (" + id + ")")
        self.callbackQueue.async {
          self.delegate?.onResponse(id: id, data: data)
        }
      }
    } else if type == "error" {
      let id = parsed["id"].stringValue
      NSLog("[SpaceX-WS]: Critical Error (" + id + "): Retrying")
    } else if type == "ping" {
      self.queue.async {
        if (self.state == .started) {
          if (self.mode == "openland") {
            self.writeToSocket(msg: JSON(["type": "pong"]))
          }
        }
      }
    }else if type == "pong" {
      self.schedulePing()
    } else if type == "complete" {
      let id = parsed["id"].stringValue
      NSLog("[SpaceX-WS]: Complete (" + id + ")")
      self.callbackQueue.async {
        self.delegate?.onCompleted(id: id)
      }
    }
  }
    
  func onClose() {
    NSLog("[SpaceX-WS]: onDisconnected")
    if self.state == .started {
      self.callbackQueue.async {
        self.delegate?.onDisconnected()
        self.delegate?.onSessiontRestart()
      }
    }
    self.stopClient()
    self.state = .waiting
    self.failuresCount += 1
    
    
    self.doConnect()
  }
  
  
  private func schedulePing() {
    if self.mode != "openland" {
      return
    }
    NSLog("[SpaceX-WS]: schedule ping")
    self.lastPingId += 1
    let pingId = self.lastPingId
    self.queue.asyncAfter(deadline: .now() + .seconds(CommonTransportLayer.PING_INTERVAL)) {
      if (self.state == .started) {
        NSLog("[SpaceX-WS]: sending ping")
        self.writeToSocket(msg: JSON(["type": "ping"]))
        self.queue.asyncAfter(deadline: .now() + .seconds(CommonTransportLayer.PING_TIMEOUT)) {
          if(self.state == .started && self.lastPingId == pingId) {
            NSLog("[SpaceX-WS]: ping timeout")
            self.onClose()
          }
        }
      }
    }
  }
  
  func close() {
    queue.async {
      NSLog("[SpaceX-WS]: Stopping")
      if self.state != .completed {
        self.state = .completed
        
        // Remove all pending requests
        self.pending.removeAll()
        
        // Stopping ws connection
        self.stopClient()
        
        // Stopping reachability
        // self.reachability.stopNotifier()
      }
    }
  }
  
  private func stopClient() {
    // Removing client if present
    let ws = self.connection
    self.connection = nil
    
    // Stopping client
    ws?.delegate = nil
    ws?.close()
  }
  
  private func writeToSocket(msg: JSON) {
    let txt = serializeJson(json: msg)
    NSLog("[SpaceX-WS]: >>" + msg["type"].stringValue)
    self.connection!.send(message: txt)
  }
}
